tests/test_tomography_gate_set_tomography.py

```diff
@@ -0,0 +1,516 @@
+
+
+@pytest.mark.parametrize("k, nqubits", [(0, 3), (1, 4), (2, 5), (3, 6)])
+def test_prepare_states_valid_k_invalid_nqubits(k, nqubits):
+    # Test for value input with invalid nqubits
+
+    # Check if ValueError is raised
+    with pytest.raises(ValueError, match="nqubits needs to be either 1 or 2"):
+        circuit = prepare_states(k, nqubits)
+
+
+@pytest.mark.parametrize("k, nqubits", [(17, 1), (18, 1), (21, 2), (56, 2)])
+def test_prepare_states_invalid_k_valid_nqubits(k, nqubits):
+    # Check if IndexError is raised
+    with pytest.raises(IndexError):
+        circuit = prepare_states(k, nqubits)
+
+
+# ##################################################################################
+
+
+@pytest.mark.parametrize("j", np.arange(0, 4, 1))
+def test_measurement_basis_value_j_single_qubit(j):
+    # Test for valid input with a single qubit
+    nqubits = 1
+
+    test_circuit = qibo.models.Circuit(nqubits)
+    new_circuit = measurement_basis(j, test_circuit)
+
+    assert isinstance(new_circuit, qibo.models.circuit.Circuit)
+    assert new_circuit.nqubits == nqubits
+
+
+@pytest.mark.parametrize("j", np.arange(0, 16, 1))
+def test_measurement_basis_value_j_two_qubits(j):
+    # Test for valid input with two qubits
+    nqubits = 2
+
+    test_circuit = qibo.models.Circuit(nqubits)
+    new_circuit = measurement_basis(j, test_circuit)
+
+    assert isinstance(new_circuit, qibo.models.circuit.Circuit)
+    assert new_circuit.nqubits == nqubits
+
+
+@pytest.mark.parametrize(
+    "j, nqubits",
+    [
+        (0, 3),
+        (1, 4),
+        (2, 5),
+        (3, 3),
+        (4, 6),
+        (5, 8),
+        (6, 7),
+        (7, 9),
+        (8, 5),
+        (9, 6),
+        (10, 4),
+        (11, 3),
+        (12, 17),
+        (13, 12),
+        (14, 4),
+        (15, 5),
+    ],
+)
+def test_measurement_basis_valid_j_invalid_nqubits(j, nqubits):
+    # Test for valid input with invalid qubits
+    test_circuit = qibo.models.Circuit(nqubits)
+
+    # Check if ValueError is raised
+    with pytest.raises(ValueError, match="nqubits needs to be either 1 or 2"):
+        new_circuit = measurement_basis(j, test_circuit)
+
+
+@pytest.mark.parametrize("j, nqubits", [(4, 1), (8, 1), (17, 2), (21, 2)])
+def test_measurement_basis_invalid_j_valid_nqubits(j, nqubits):
+    # Test for invalid input with valid nqubits
+    test_circuit = qibo.models.Circuit(nqubits)
+
+    # Check if IndexError is raised
+    with pytest.raises(IndexError):
+        new_circuit = measurement_basis(j, test_circuit)
+
+
+# ##################################################################################
+
+
+def test_reset_register_valid_string_1qb():
+    # Test for valid string
+    nqubits = 1
+    test_circuit = qibo.models.Circuit(nqubits)
+    test_circuit.add(gates.H(0))
+
+    invert_register = "sp_0"
+    inverted_circuit = reset_register(test_circuit, invert_register)
+
+    assert isinstance(inverted_circuit, qibo.models.circuit.Circuit)
```

Comment:
Same comment as above, test for the expected outputs. This applies for all the other tests below, when you can try to unify them using pytest parameters, I believe some improvement is still possible.